src/hyperlight_host/src/sandbox/mod.rs

```diff
@@ -86,6 +93,102 @@ pub fn is_hypervisor_present() -> bool {
+    /// The start guest time, in TSC cycles, for the current guest.
+    #[allow(dead_code)]
+    pub guest_start_tsc: Option<u64>,
+    /// The file to which the trace is being written
+    #[allow(dead_code)]
+    pub file: Arc<Mutex<std::fs::File>>,
+    /// The unwind information for the current guest
+    #[cfg(feature = "unwind_guest")]
+    #[allow(dead_code)]
+    pub unwind_module: Arc<dyn crate::mem::exe::UnwindInfo>,
+    /// The framehop unwinder for the current guest
+    #[cfg(feature = "unwind_guest")]
+    pub unwinder: framehop::x86_64::UnwinderX86_64<Vec<u8>>,
+    /// The framehop cache
+    #[cfg(feature = "unwind_guest")]
+    pub unwind_cache: Arc<Mutex<framehop::x86_64::CacheX86_64>>,
+}
+#[cfg(feature = "trace_guest")]
+impl TraceInfo {
+    /// Create a new TraceInfo by saving the current time as the epoch
+    /// and generating a random filename.
+    pub fn new(
+        #[cfg(feature = "unwind_guest")] unwind_module: Arc<dyn crate::mem::exe::UnwindInfo>,
+    ) -> crate::Result<Self> {
+        let mut path = std::env::current_dir()?;
+        path.push("trace");
+        path.push(uuid::Uuid::new_v4().to_string());
+        path.set_extension("trace");
+        #[cfg(feature = "unwind_guest")]
+        let hash = unwind_module.hash();
+        #[cfg(feature = "unwind_guest")]
+        let (unwinder, unwind_cache) = {
+            let mut unwinder = framehop::x86_64::UnwinderX86_64::new();
+            unwinder.add_module(unwind_module.clone().as_module());
+            let cache = framehop::x86_64::CacheX86_64::new();
+            (unwinder, Arc::new(Mutex::new(cache)))
+        };
+        let tsc_freq = Self::calculate_tsc_freq()?;
+
+        let ret = Self {
+            epoch: std::time::Instant::now(),
+            tsc_freq,
+            guest_start_epoch: None,
+            guest_start_tsc: None,
+            file: Arc::new(Mutex::new(std::fs::File::create_new(path)?)),
+            #[cfg(feature = "unwind_guest")]
+            unwind_module,
+            #[cfg(feature = "unwind_guest")]
+            unwinder,
+            #[cfg(feature = "unwind_guest")]
+            unwind_cache,
+        };
+        /* write a frame identifying the binary */
+        #[cfg(feature = "unwind_guest")]
+        self::outb::record_trace_frame(&ret, 0, |f| {
+            let _ = f.write_all(hash.as_bytes());
+        })?;
+        Ok(ret)
+    }
+
+    /// Calculate the TSC frequency based on the RDTSC instruction.
+    fn calculate_tsc_freq() -> crate::Result<u64> {
+        if !hyperlight_guest_tracing::invariant_tsc::has_invariant_tsc() {
+            return Err(crate::new_error!(
+                "Invariant TSC is not supported on this platform"
+            ));
+        }
+        let start = hyperlight_guest_tracing::invariant_tsc::read_tsc();
+        let start_time = std::time::Instant::now();
+        // Sleep for 1 second to get a good sample
+        std::thread::sleep(std::time::Duration::from_secs(1));
```

Comment:
Will this cause a delay when running with tracing feature?

Reply:
Removed this delay. Now the frequency is calculated using two readings of TSC on the host.
One when the guest is started, the other when the first trace records are received.